File: TEO/Mav.UI/Mav.Ui/src/app/services/i-user-assessment.service.ts
import { Observable } from 'rxjs/Observable';
import { GenericResult } from '../models/generic.result.model';
import { Paginate } from '../models/paginated.items';
import { UserAssessmentSearchModel } from '../models/search/user.assessment.search.model';
import { UserAssessmentListingModel } from '../models/search/user.assessment.listing.model';
import { AddUserAssessmentViewModel } from '../models/add.user.assessment.view.model';
import { UpdateUserAssessmentViewModel } from '../models/update.user.assessment.view.model';
import { SubmitUserAssessmentViewModel } from '../models/submit.user.assessment.view.model';

export interface IUserAssessmentService {

    getUserAssessmentById<T>(id: number): Observable<T>;

    getUserAssessmentListByUserId(filter: UserAssessmentSearchModel): Observable<Paginate<UserAssessmentListingModel>>;

    getUserLatestAssessment(id: string): Observable<UserAssessmentListingModel>;

    getUserHistory(filter: UserAssessmentSearchModel): Observable<Paginate<UserAssessmentListingModel>>;

    createUserAssessment(newItem: AddUserAssessmentViewModel): Observable<GenericResult<number>>;

    updateUserAssessment(item: UpdateUserAssessmentViewModel): Observable<GenericResult<number>>;

    submitUserAssessment(item: SubmitUserAssessmentViewModel): Observable<GenericResult<number>>;

}

File: TEO/Mav.Services/Abstract/IUserAssessmentService.cs
using Mav.Common.Models;
using Mav.Data.Repositories;
using Mav.Models.SearchModels;
using Mav.Models.ViewModels;
using System.Threading.Tasks;

namespace Mav.Services.Abstract
{
    public interface IUserAssessmentService
    {
        Task<T> GetById<T>(int id);

        Task<IPaginate<UserAssessmentListingModel>> GetListByUserId(UserAssessmentSearchModel vm);

        Task<UserAssessmentListingModel> GetUserLatestAssessment(string id);

        Task<IPaginate<UserAssessmentListingModel>> GetUserHistory(UserAssessmentSearchModel vm);

        Task<GenericResult<T>> Add<T>(AddUserAssessmentViewModel vm);

        Task<GenericResult<T>> Update<T>(UpdateUserAssessmentViewModel vm);

        Task<GenericResult<T>> Submit<T>(SubmitUserAssessmentViewModel vm, InvigilatorViewModel userModel);

    }
}
